sql/plan/cached_results.go

```diff
@@ -69,6 +93,7 @@ func (n *CachedResults) RowIter(ctx *sql.Context, r sql.Row) (sql.RowIter, error
 }
 
 func (n *CachedResults) Dispose() {
+	n.cacheDone = false
```

Comment:
This is kind of non-sensical, smells wrong to me.

Dispose should happen at the end of a node's lifecycle, when all rows have been iterated and there is no chance that additional work is going to be performed on it. The fact that you're doing this early from HashLookup doesn't change this equation.

Shouldn't we be setting n.cacheDone when the child iter return EOF? Is it actually cacheInitialized?

Reply:
is this just a naming confusion? to me `cacheInitialized` sounds the cache has started being filled, when i intend it to mean "the cache contains all of the rows of an exhausted child iter". I can make it act like the former and maintain its usefulness. The point of this line is to help make the parent idiot-proof, even if the parent does something not sensible this prevents the infinite spinning bug. the CachedResult does not control whether its parent does a sensible thing.

in aggregate i don't think this complexity is necessary, i think we should delete CachedResult and nodes that want caches should just build them in-line